ATM: Implement the current endpoint filters as EndpointCharacteristics

tiferet opened the pull request:
📢 This PR is a bit bigger, so commit-by-commit review might be easier. I've tried to make the commit comments as informative as possible.

### Main changes
- Implement both the standard and the type-specific endpoint filters as `EndpointCharacteristic`s.
- Move the definitions of `isEffectiveSink` and `getAReasonSinkExcluded` to the base class, as they can now be implemented generically for all sink types. 
- Changing the definition of `getAReasonSinkExcluded` is how we'd adjust which endpoints we score at inference time. For now I've implemented it to replicate the logic in the old code, so that results remain unaffected. I've tracked possible experiments to improve this selection in https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2126.

### A few notes

Note that this PR still sticks to the principle of not breaking any tests, _except_ that I had to disambiguate three filters from three different sink types that all had the same name (fc56c5a022bf56e962f7a9052313f253dcee603d), and that required a tiny update to `FilteredTruePositives.expected` (0fd013f9fd83d74e0fb443708c7d999957fda99c).

Also note that the training data is unaffected because (for now) I've given all EndpointFilterCharacteristics medium confidence, whereas only high-confidence characteristics contribute to training set selection. AIUI, the reason endpoint filters weren't used to select negative training samples in the old code was precisely this: their accuracy is high enough that we don't want to waste inference time scoring these endpoints, but not high enough that we can reliably use them as negative training samples. It's worth having someone with the needed expertise (Stephan? 😉) go through them eventually to consider whether any should be promoted to high confidence. I tracked this possible experiment in https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2126.

### Timing checks
- ✅ KPI timing experiment: https://github.com/github/codeql-dca-main/issues/8634
- ☑️ The local runtime of `endpoint_large_scale/ExtractEndpointDataTraining` remains like it was after the [last PR](https://github.com/github/codeql/pull/11263): About 5s.


Closes https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2100

[Probably](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2101#issuecomment-1317628964) also closes https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2101?

adityasharad commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:
```diff
@@ -69,7 +69,26 @@ abstract class AtmConfig extends string {
+      specificFilter.getEndpoints(candidateSink) and
+      specificFilter.getImplications(getASinkEndpointType(), false, _) and
+      result = specificFilter
+    )
```
This pattern of having to downcast `result` to two possible subtypes of `EndpointCharacteristic` is a little unfortunate (in any object-oriented language), and suggests we should be making more use of the ability to define predicates within the object-oriented type hierarchy.

Ideally you would just say something like:
```ql
result.isEndpointForType(candidateSink, this.getASinkEndpointType())
```
and define/override that predicate correctly on both `EndpointFilterCharacteristic` and `StandardEndpointFilterCharacteristic`.

This change would be a bit invasive, so let's come back to this one separately.

tiferet replied:
This is just an artifact of wanting to perfectly reproduce the existing data for now, while building a class structure that makes sense for the long term. In the long term I envision `getAReasonSinkExcluded` simply saying that we exclude endpoints whose class is known with medium-high confidence a priory (ie, we only score Unknown endpoints). That can be written just by querying the implications across all EndpointCharacteristics. See the second checklist item [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2126#issue-1450274932).

adityasharad replied:
Sounds good.

adityasharad reviewed (commented):
Nice. I can see the existing filtering logic translated into the new object-oriented hierarchy, and that you've taken care to preserve that logic.

My suggestions are mostly about the clarity and performance of the QL code, rather than the semantics of the filters. Most are not blocking but I hope will be useful.

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:
```diff
@@ -442,3 +446,112 @@ private class BuiltinCallNameCharacteristic extends ArgumentToBuiltinFunctionCha
+  override predicate getEndpoints(DataFlow::Node n) {
+    // Ignore candidate sinks within externs, generated, library, and test code
+    ClassifyFiles::classify(n.getFile(), category) and
+    this = "in " + category + " file"
```
I din't remember why I put it there 😄 It does seem redundant, though -- thanks. Deleting.

tiferet replied:
Done in 8d22fd25f1b12f3cb8bdd37f302ebd2b79a23e61

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/StandardEndpointFilters.qll:
```diff
@@ -10,25 +10,13 @@ private import javascript
+  exists(EndpointCharacteristics::StandardEndpointFilterCharacteristic characteristic |
+    characteristic.getEndpoints(n) and
+    result = characteristic
   )
```
Yeah, I've used that in some places, but wasn't sure which one to prefer. Does one have any advantages over the other?

(This file gets deleted altogether in a later PR, so asking for my general knowledge.)

adityasharad replied:
Equivalent in meaning and performance. The key difference is that:
- [`any`](https://codeql.github.com/docs/ql-language-reference/expressions/#any) is an expression, so it has a "return" value, which is convenient when you need to use the value of the variables outside the quantifier.
- [`exists`](https://codeql.github.com/docs/ql-language-reference/formulas/#exists) is a formula, so it is convenient when you need to say something about the quantified variables but don't need their value outside the quantifier.

tiferet replied:
Ah, I see, thanks! Good to know!

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:
```diff
@@ -555,3 +556,311 @@ private class InIrrelevantFileCharacteristic extends StandardEndpointFilterChara
+  override predicate getEndpoints(DataFlow::Node n) {
+    exists(DataFlow::CallNode call | n = call.getAnArgument() |
+      // additional databases accesses that aren't modeled yet
+      call.(DataFlow::MethodCallNode).getMethodName() =
```
Nope, this is just copy-pasted directly from the old endpoint filters -- I haven't read through each of them. Yes, I'll make that change 👍

tiferet replied:
Done in 8d22fd25f1b12f3cb8bdd37f302ebd2b79a23e61

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:
```diff
@@ -555,3 +556,311 @@ private class InIrrelevantFileCharacteristic extends StandardEndpointFilterChara
+  }
+}
+
+private class NotDirectArgumentToLikelyExternalLibraryCallOrHeuristicSinkNosqlCharacteristic extends NosqlInjectionSinkEndpointFilterCharacteristic {
```
Yeah, I wrote a comment about this [here](https://github.com/github/codeql/pull/11321#discussion_r1025681567)

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:
```diff
@@ -555,3 +556,311 @@ private class InIrrelevantFileCharacteristic extends StandardEndpointFilterChara
+    // `codeql/javascript/ql/src/semmle/javascript/heuristics/AdditionalSinks.qll`.
+    // We can't reuse the class because importing that file would cause us to treat these
+    // heuristic sinks as known sinks.
+    not n = StandardEndpointFilters::getALikelyExternalLibraryCall().getAnArgument() and
+    not (
+      isAssignedToOrConcatenatedWith(n, "(?i)(nosql|query)") or
+      isArgTo(n, "(?i)(query)")
+    )
```
This again is copied directly from the original endpoint filters. I'd rather let @kaeluka decide about improvements to them, because ultimately he's going to be their owner. Though it's possible you're suggesting a trivial change, and I'm just misunderstanding you?

adityasharad replied:
Ah I see. Let's think about these separately then; I think there's room for improvement but don't have an easy fix that wouldn't change the meaning, and you didn't introduce this in the PR.

tiferet replied:
Tracked [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2136)

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:
```diff
@@ -442,3 +446,112 @@ private class BuiltinCallNameCharacteristic extends ArgumentToBuiltinFunctionCha
+private class IsArgumentToModeledFunctionCharacteristic extends StandardEndpointFilterCharacteristic {
+  IsArgumentToModeledFunctionCharacteristic() { this = "argument to modeled function" }
+
+  override predicate getEndpoints(DataFlow::Node n) {
```
Good shoutout! Indeed this is a big change to make in an intermediate PR, but I'll do it in the last PR of the series. 

I think `hasEndpoints` might make more sense than `isEndpoint`. Really, it's the endpoint that has this characteristic, not the characteristic that has the endpoint. Logically I would prefer `appliesToEndpoint`, but that's not in line with the style guide 🤷

adityasharad replied:
`appliesToEndpoint` is even better! Go with that.

tiferet replied:
Event though it doesn't fit the style guide?

tiferet replied:
Tracked [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2136)

adityasharad replied:
Absolutely. My intention is not to strictly enforce the style guide, but to have meaningful predicate names that help us understand the intent and the expected result type. This is a good name.

Another good naming trick is to decide how you would [write a doc comment for the predicate](https://github.com/github/codeql/blob/main/docs/qldoc-style-guide.md#qldoc-for-predicates), and whether your predicate name aligns with that sentence e.g. I would document this predicate as "Holds if `this` characteristic applies to the endpoint node `n`."

tiferet replied:
👍 I wasn't sure how strict we are about the style guide in this repo

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:
```diff
@@ -442,3 +446,112 @@ private class BuiltinCallNameCharacteristic extends ArgumentToBuiltinFunctionCha
+  bindingset[this]
+  StandardEndpointFilterCharacteristic() { any() }
+
+  override predicate getImplications(
```
👍 I'll do this to in a followup PR.

tiferet replied:
Tracked [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2136)

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:
```diff
@@ -555,3 +556,311 @@ private class InIrrelevantFileCharacteristic extends StandardEndpointFilterChara
+    // `codeql/javascript/ql/src/semmle/javascript/heuristics/AdditionalSinks.qll`.
+    // We can't reuse the class because importing that file would cause us to treat these
+    // heuristic sinks as known sinks.
+    not StandardEndpointFilters::flowsToArgumentOfLikelyExternalLibraryCall(n) and
```
Tracked [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2136)

kaeluka reviewed (changes requested):
I love love love how this is remolding logic that used to be really hard to manage into something more maintainable. Things are really starting to fall into place! Thank you :)

The one complaint I have about this is how the `StandardEndpointFilterCharacteristic` class is being made magic by its usage in other modules. The bright side is that there's very easy fixes that I have outlined in my comments. Addressing these should be doable in 20-30min.

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/EndpointCharacteristics.qll:
```diff
@@ -442,3 +447,418 @@ private class BuiltinCallNameCharacteristic extends ArgumentToBuiltinFunctionCha
+  ) {
+    endpointClass instanceof NegativeType and
+    isPositiveIndicator = true and
+    confidence = mediumConfidence()
```
AIUI, the current thinking is that we add to the training set only endpoints whose class we're fairly sure about (so as not to train on noisy data). The endpoint filters, OTOH, don't need to be perfect: we're OK with losing some potential TPs to speed up inference and avoid lots of FPs. My understanding is that the reason these were used only at inference time and not for training data extraction _is_ that they have lower confidence.

The fact that this was the historical reasoning doesn't mean we need to keep it forever: I'm all for experimenting with changing confidence where you think it's appropriate. But again, this first round of PRs is meant to replicate the existing data exactly. I'm tracking data improvements we can experiment with later [here](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2126#issue-1450274932). In fact, I think the question you raised here is tracked for experimentation as the third item in that list.

kaeluka replied:
OK, it's certainly acceptable to skip discussing what the perfect value here should be. Resolving!

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:
```diff
@@ -73,6 +73,10 @@ query predicate trainingEndpoints(
       c instanceof LikelyNotASinkCharacteristic
     )
   ) and
+  // Don't surface endpoint filters as characteristics, because they were previously not surfaced.
+  // TODO: Experiment with surfacing these to the modeling code by removing the following line (and then make
+  // EndpointFilterCharacteristic private).
+  not characteristic instanceof EndpointFilterCharacteristic and
```
See the TODO comment directly above this line. And also see [this comment](https://github.com/github/codeql/pull/11321#discussion_r1025681567).

kaeluka replied:
👍: I noticed that ATMConfig.qll also depends on `EndpointFilterCharacteristic` which it needs to do in order to not change behaviour. This is disappointing, but not the fault of the PR. This pushes the level of work required for a fix to a point where I think it's best to stick with that current implementation for now and tackle this during experimentation.

Therefore, I'm closing this thread!

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/lib/experimental/adaptivethreatmodeling/ATMConfig.qll:
```diff
@@ -69,7 +69,28 @@ abstract class AtmConfig extends string {
+    // characteristics that are specific to this sink type.
+    // TODO: Experiment with excluding all endpoints that have a medium- or high-confidence characteristic that implies
+    // they're not sinks for this sink type (or not sinks for any sink type), not just the EndpointFilterCharacteristics.
+    exists(EndpointCharacteristics::StandardEndpointFilterCharacteristic standardFilter |
```
Again, see the TODO comment right above this. I don't know if EndpointFilterCharacteristics will even exist in the long run (I hope not). They are a way of replicating the existing logic with the new framework. In the long run all selection (training and inference) should ideally depend only on the implications of the characteristics, and not on their class structure. If being an EndpointFilterCharacteristic doesn't carry any particular implications, then the selection code shouldn't know or care that this class exists.

tiferet replied:
Again see my TODO comment directly above this. Same thing here as all the other places. I believe that's the right approach, but it needs to be verified experimentally.

kaeluka replied:
A TODO would be the appropriate choice if the fix would take a lot of work; but the fix here is a matter of minutes.

If you're currently juggling too many things at the same time, I'd be more than happy to help out by adding a small commit first thing tomorrow morning! :) Just let me know in that case.

tiferet replied:
Making this change requires end-to-end testing to ensure it doesn't negatively impact our metrics. That's far from trivial, though it will be easier once we have the MVP of the orchestrator. Like we agreed, I'm first making the transition to the new framework _without_ impacting the data we produced, and then we can start doing all the experiments we want to improve the data.

kaeluka replied:
> Making this change requires end-to-end testing to ensure it doesn't negatively impact our metrics.

That's not true; I just redid what I did yesterday locally. Using the clause:

```
exists(EndpointCharacteristics::EndpointCharacteristic filter, float c |
      filter.getEndpoints(candidateSink) and
      c >= any(EndpointCharacteristics::EndpointCharacteristic ch).mediumConfidence() and
      filter.getImplications(any(NegativeType n), true, c) and
      result = standardFilter
    )
```

filters exactly the same endpoints as your clause, but in a way that is more modular. If it would've changed the data selection, I'd not have (knowingly) suggested a change.

> Incidentally, it wouldn't be c >= any(EndpointCharacteristic c).maximalConfidence(): endpoint filters have medium confidence, not maximal confidence.

I know, I had mentioned this in the first comment of this thread.

tiferet replied:
Endpoint filters are _not_ maximal confidence -- quite the opposite, they're lower confidence than NotASink. AIUI, your commit should have added lots of negative examples to the training set that weren't there before -- I'm not sure why it didn't fail `endpoint_large_scale/ExtractEndpointDataTraining.qlref`. At the moment there isn't a test that verifies the selection of endpoints for inference (see https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2135), so that needs to be verified via a DCA experiment. 

In any case, I could definitely achieve something like this by carefully setting and then filtering on confidence levels, to ensure that in practice the training set chooses only the historical NotASink endpoints and inference runs only on the historical effective sink endpoints. I feel this would hide the fact that these current choices are fragile and often arbitrary, though. I prefer to be explicit about the fact that we have some arbitrary selection going on at the moment, by explicitly using subclasses for selection, and then change that ugliness when we're truly rationalizing the selection by setting proper confidence levels and using them consistently.

kaeluka replied:
TL;DR 🤝 : I think we may have found a compromise.

At this point, I want to explain why I care about this so much; it may seem like an irrelevant detail, but to me it's far from it.

The framework we're building is based on expressing our believes about classes of endpoints in terms of implications, which are labels involving a sink type, a boolean, and a confidence. This is a really simple framework (I mean 'simple' in a good way). However, if this framework would turn out to be TOO simple to express what we need to express, we're painting ourselves in a corner: That's why I react strongly to all cases where selection of endpoints is not based on the implications, but on the specific characteristic class. in these cases, the implications framework was not sufficient to express what we need to express (or it was too hard). Saying that these situations will all be resolved later on in the process (as the TODOs do) can't be done with certainty. In order to remove those cases, we'll need experiments to show that resolving the situations does not make the models worse. If the odd legacy rules happen to be weird, but work in practice, we'd be stuck with that strange logic for a long time, or forever.

As we have decided to go with a process that splits the refactoring into many small PRs, it's extra important to address these concerns earlier, rather than later — later, the resistance to undo so much work (in the worst case) would be huge.

> Endpoint filters are not maximal confidence -- quite the opposite, they're lower confidence than NotASink

I'm totally with you there - I don't care about using this specific value at all, any value that'll reproduce the current behaviour will suffice. *BUT*, I'd say that this assignment of values looks so strange is a *feature* of the new framework: it makes this weirdly high confidence setting stick out like a sore thumb.

> In any case, I could definitely achieve something like this by carefully setting and then filtering on confidence levels, to ensure that in practice the training set chooses only the historical NotASink endpoints and inference runs only on the historical effective sink endpoints. I feel this would hide the fact that these current choices are fragile and often arbitrary, though.

These choices, on SOME level will always be arbitrary, as we only merge settings that are backed by experimental evidence. That being said, we both are committed to finding better values for them in the future, which means it will definitely happen.

My current worry, though, is about expressivity of the framework, and in the future it may be too late to make changes.

If you'd be willing to use any values that make it all work, we'd have a compromise that I could feel almost relaxed about (you know me, I'm never *really* relaxed, so that's as good as it gets ^^), merge this PR and move on to the next.

If you want extra assurance that adapting the confidence levels gets tackled later: I can make an issue and assign it to myself so we don't lose track of that promise. Or if you'd rather own that, we can assign it to you - but I'm offering to do the work, as I think this might get us to the handshake.

tiferet replied:
OK, thank you for clarifying -- I have a much better understanding of your concerns now 🙏 I suspect we'll still have some back-and-forth on this, but we're getting closer to understanding one another's concerns, which is a win!

My guess is we won't be _entirely_ able to replace all the current logic using only confidence levels, because at some point we'll hit something where logically we think a characteristic has confidence X but in practice, experimentally we get bad results when we do that. And that's OK -- a rule is still useful even when it has a small number of exceptions: the key is to make those exceptions as clear and visible as possible, and make the entire codebase easy to understand and maintain. In my mind there's no question that this new framework accomplishes that, even if we won't be able to remove every single edge case we're currently reproducing from the old code. It'll be very easy to see what the remaining edge cases are and why they're there, which was _not_ easy to see with the old framework. The edge cases will be encoded in exactly two places: edge cases for the training set will be encoded explicitly in the [predicate that selects endpoints for training](https://github.com/github/codeql/blob/3fbe089f65c291030389ef41ba995a5506783346/javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll#L53), and edge cases for the inference set will be encoded explicitly in the predicate that selects endpoints for inference (which is this one). Setting weird confidence levels hides these exceptions. Calling the exceptions out directly in the extraction predicate makes them explicit and visible.

tiferet replied:
I created [this commit](https://github.com/github/codeql/commit/03b8e649f1528905ed1709ce4581e9368337cccf) in a separate branch to show how easily we can replicate the current selection using only confidence levels. This is exactly what I think we want to do long-term, _except for_ line 86: `confidence < filter.highConfidence() and`. Logically it makes no sense to score endpoints that we are highly certain are _not_ sinks at inference time. The reason this line is needed is that there are currently three high-confidence (NotASink) characteristics that are shared by many TP sinks: 
- JQueryArgument (Xss sinks)
- FileSystemAccess (tainted path sinks)
- DatabaseAccess (NoSQL / SQL sinks)

When I delete that line, `endpoint_large_scale/FilteredTruePositives.ql` fails, with lots of newly-filtered TPs, which could well impact ATM metrics. Of course, `endpoint_large_scale/FilteredTruePositives.ql` already has many exceptions hard-coded in, which means some of the current endpoint filters also apply to a decent number of TPs. I don't know whether they were added to speed up inference or to improve ATM metrics, but I would hope we could surface more TPs by removing all these endpoint filters. We should certainly consider removing `JQueryArgument`, `FileSystemAccess`, and `DatabaseAccess` from the set of high-confidence characteristics, because they may be adding lots of bad negative examples to the training set.

Merging this commit would hide all these issues, and thereby reduce our ability to make adjustments that might improve ATM metrics considerably. At the moment, we have a set of decisions that were made for unknown reasons at some point in history. Some of them might indeed be critical to reaching good metrics and/or good runtime, but others are likely doing more harm than good. Only experimentation will tell us which of these decisions we can undo and which ones we have to keep. Of the ones we keep, preferably we can encode them by properly adjusting confidence levels, but there may be some corner cases where, for whatever reason, we want to, say include some type of endpoint in the negative training set without excluding it as a candidate for scoring 🤷

tiferet replied:
Having said all that, if you feel very strongly about it, I would be willing to merge [this commit](https://github.com/github/codeql/commit/03b8e649f1528905ed1709ce4581e9368337cccf) (with some updates to the documentation), because it still upholds the rule that edge cases are encoded only in the endpoint extraction queries. In that sense, it still makes it easy to find all the edge cases, keeping them explicit and visible. What it doesn't preserve is our ability to find all edge cases simply by making all EndpointCharacteristic sub-classes private, but if you feel strongly about this that's a price I'm willing to pay 🤷‍♂️

kaeluka replied:
OK, then we have a deal! Yay! 🥳 This makes me happy to hear 👍

I'm still a bit concerned about how the confidences framework will work out in the end, but I thought lots about it last night and there's a plan B that should cover us in the very very worst case. So I think that merging with this commit is the reasonable thing to do.

kaeluka commented:
I'm still working on this review, trying to figure out what the *real* problem is here and coming up with a specific solution. The PR's interdependencies are quite complex, which is why this has taken a little.

kaeluka commented:
I think I'm through with the review.

🤡 The PR sadly reproduces some of the confusing definitions we currently have in `main`. How the classes `StandardEndpointFilterCharacteristic` and `EndpointFilterCharacteristic` are used is quite confusing. This isn't a blocker though, as these classes will probably change LOTS during future experimentation. This is why I think we can safely ignore this paper cut for now and should focus on getting things done.

I'm a bit sick and will call it a day now.

✅ I consider the review of this PR done and approved, assuming you are making the `StandardEndpointFilterCharacteristic` private (for example, by adding the commit I linked above).
✅ I reviewed the PR that's [removing a lot of files](https://github.com/github/codeql/pull/11321) (😍) and all it'd need to be merged is a rebase.
🕥 Tomorrow, I'll work on reviewing the third PR (https://github.com/github/codeql/pull/11323) but to manage your expectations: that one sounds complex and might take a while to fully grok.

tiferet commented:
(Minor request: Do you mind using words rather than emoji? I don't remember off the top of my head what each emoji was meant to signify, and I had to spend some time searching for the slack thread to look up what 🤡 means)

tiferet commented on javascript/ql/experimental/adaptivethreatmodeling/modelbuilding/extraction/ExtractEndpointDataTraining.qll:
```diff
@@ -73,6 +73,10 @@ query predicate trainingEndpoints(
+  // Don't surface endpoint filters as characteristics, because they were previously not surfaced.
+  // TODO: Experiment with surfacing these to the modeling code by removing the following line (and then make
+  // EndpointFilterCharacteristic private).
+  not characteristic instanceof EndpointFilterCharacteristic and
```
OK, I see why your commit didn't fail `endpoint_large_scale/ExtractEndpointDataTraining.qlref`: It's because it didn't remove this line. If we were to raise the confidence of the endpoint filters without having this line, the training set would include all non-effective-sinks, which would be bad, because many of those are in fact low-confidence characteristics. 

Do you see a point in making `StandardEndpointFilterCharacteristic` private without removing this line or making `EndpointFilterCharacteristic` private?

IMO, I think we need to be consistent: Either we use subclasses of EndpointCharacteristics to replicate the existing data for now, or we try to improve the logic at the same time that we improve the code, and each PR requires end-to-end testing. I definitely prefer the former. The nice thing is that, once we're ready to progress to the second phase, even without my `TODO` comments and [issue tracking](https://github.com/github/ml-ql-adaptive-threat-modeling/issues/2126), we can easily look at all EndpointCharacteristics that aren't private and find the places they're used in the code, and that will tell use where we want to make improvements. Ultimately all EndpointCharacteristics should be private, with subclasses used only for two things: (1) Abstract classes are used to apply the same set of implications to many different subclasses without needing to overwrite `getImplications()` over and over. (2) Non-abstract classes are used to define the labels for type balancing through `getEndpoints()`.

tiferet commented:
> ✅ I consider the review of this PR done and approved, assuming you are making the `StandardEndpointFilterCharacteristic` private (for example, by adding the commit I linked above). 

Unfortunately I don't think I can do that 😞. Or rather, I _can_ do it by introducing a new confidence level, but that would further hide ugliness in a way that will make it harder to fix later. I know this back-and-forth is frustrating, and I'm sure a quick sync would be an easier way to reach resolution, but we can do that when you're feeling better -- please don't push yourself!

> ✅ I reviewed the PR that's [removing a lot of files](https://github.com/github/codeql/pull/11321) (😍) and all it'd need to be merged is a rebase. 

Thank you! ❤️ 

> 🕥 Tomorrow, I'll work on reviewing the third PR (#11323) but to manage your expectations: that one sounds complex and might take a while to fully grok.

No rush! [There's a chance that one is trickier than I realized](https://github.slack.com/archives/CP9127VUK/p1669045505695319). Luckily, if we do need big changes on that one, there's only [one PR](https://github.com/github/codeql/pull/11333) following it in the chain, so it won't require big changes to a long chain of subsequent PRs. Anyway, Thursday and Friday are bank holidays in the US, so unless that PR is unexpectedly ready for approval and merging tomorrow, it won't get merged before Monday. More importantly, get lots of rest and get healthy!

kaeluka commented:
> More importantly, get lots of rest and get healthy!

Thank you! I agree, there's no point rushing this. I see several ways forward, we only need to find one that works for both of us. And enjoy your time off!

tiferet commented:
@kaeluka I merged [this change](https://github.com/github/codeql/pull/11462) we agreed on into this PR, and updated the predicate's documentation accordingly. I _think_ this addresses all your concerns, unless I lost track of something else in the long discussion 😅 Sending this back to you for final(?) review 🏓
